_rules/name-from-content-included-in-accessible-name-2ee8b8.md

```diff
@@ -0,0 +1,192 @@
+
+**Note:** Due to the definition of [included characters](#included-characters) the whole consecutive sequence of characters in the [visible](#visible) [child text content](https://dom.spec.whatwg.org/#concept-child-text-content) of the target element has to be included character-by-character in the [accessible name](#accessible-name) to meet the expectation of this rule.
+
+## Assumptions
+
+- This rule assumes that the [visible](#visible) [text nodes](https://www.w3.org/TR/dom/#text) follow the same order as the text in the code, e.g. CSS has not been used to re-arrange pieces of text so that they are presented in a different order.
+- This rule assumes that the whole set of [visible](#visible) [text nodes](https://www.w3.org/TR/dom/#text) of the element are styled, including placement, in such a way that a sighted user will perceive them all as part of the label of the element. An example where this would not be the case, is where styling is used to split up the visible text nodes in e.g. two parts and place one part further away, or give it a different styling that makes it appear as not connected to the element it labels.
+
+## Accessibility Support
+
+There are no major accessibility support issues known for this rule.
+
+## Background
+
+- https://www.w3.org/TR/WCAG21/#label-in-name
+
+## Test cases
+
+### Passed
+
+#### Passed Example 1
+
+Visible label and accessible name matches when trailing white spaces are removed.
+
+```html
+<div role="link" aria-label="next page ">next page</div>
+```
+
+#### Passed Example 2
+
+Character insensitivity between visible label and accessible name.
+
+```html
+<div role="link" aria-label="Next Page">next page</div>
+```
+
+#### Passed Example 3
+
+Full visible label is included in the accessible name.
+
+```html
+<button name="link" aria-label="Next Page in the list">Next Page</button>
+```
+
+#### Passed Example 4
+
+Full visible label is included in the accessible name.
+
+```html
+<head>
+  <style>
+    .sr-only {
+      position:absolute;
+      left:-10000px;
+      top:auto;
+      width:1px;
+      height:1px;
+      overflow:hidden;}
+  </style>
+</head>
+
+<body>
+  <a href="home.html">Read more<span class="sr-only"> about our pricing</span></a>
+</body>
+```
+
+#### Passed Example 5
+
+Text nodes in name from content are not expressing anything in human language and do therefore not live up to the [WCAG definition of text](https://www.w3.org/TR/WCAG21/#dfn-text).
+
+```html
+<button aria-label="close">:-)</button>
+```
+
+#### Passed Example 6
+
+Text nodes in name from content are not expressing anything in human language and do therefore not live up to the [WCAG definition of text](https://www.w3.org/TR/WCAG21/#dfn-text). In this case, "X" is used because it resembles a "close" icon, not to represent the character "X".
+
+```html
+<button aria-label="close">X</button>
+```
+
```

Comment:
What about
```html
<button aria-label="close">Close :-(</button>
```
With the current expectation, this fails because not all visible text content is included in the accessible name. But what is not included is not expressing something in human language, so I believe it should pass. Is this too much of an edge case, or should we work on the expectation to consider this?

Reply:
@carlosapaduarte (and @kasperisager), do you think this change to the expectation could cover both the existing test cases + the new edge cases that have come up?

Current Expectation: 

> The [visible](#visible) [child text content](https://dom.spec.whatwg.org/#concept-child-text-content) that are [descendants](https://www.w3.org/TR/dom41/#concept-tree-descendant) in the [flat tree](https://drafts.csswg.org/css-scoping/#flat-tree) of the target element either:
> * is [included](#included-characters) in its entirety within the [accessible name](#accessible-name) of the element, or 
> * does not express anything in [human language](https://www.w3.org/TR/WCAG21/#dfn-human-language-s) and therefore does not live up to the [WCAG definition of text](https://www.w3.org/TR/WCAG21/#dfn-text).

Suggested new expectation: 

> The [visible](#visible) [child text content](https://dom.spec.whatwg.org/#concept-child-text-content) that are [descendants](https://www.w3.org/TR/dom41/#concept-tree-descendant) in the [flat tree](https://drafts.csswg.org/css-scoping/#flat-tree) of the target element is [included](#included-characters) in its entirety within the [accessible name](#accessible-name) of the element, except for parts that do not express anything in [human language](https://www.w3.org/TR/WCAG21/#dfn-human-language-s) and therefore does not live up to the [WCAG definition of text](https://www.w3.org/TR/WCAG21/#dfn-text).

Reply:
Another question: How about punctuation and special characters? Does that have to be the same 1:1? 

E.g.:
`<a href="#" aria-label="Press me">Press me!!!!!</a>`

... because in the edge case listed further up, only the "X" can be said to be non-text content, but the parenthesis should not be included in the match either: 
`<a href="#">Close (X) and go back</a>`

Reply:
Expectation changed to the suggestion above to allow for content that is a mix between text and content that does not express anything in human language. 
Punctuation now is explicitly not considered for the comparison due to an update of [the Understanding 2.5.3 Name in Label](https://www.w3.org/WAI/WCAG21/Understanding/label-in-name.html) document.

Reply:
Test cases added for punctuation (Passed Example 6) and mixed text/non-text (Passed Example 7).